src/databricks/labs/ucx/hive_metastore/table_migrate.py

```diff
@@ -90,38 +98,103 @@ def for_cli(cls, ws: WorkspaceClient, product='ucx'):
+        all_principal_grants = None if acl_strategy is None else self._principal_grants.get_interactive_cluster_grants()
         self._init_seen_tables()
         tables_to_migrate = self._tm.get_tables_to_migrate(self._tc)
-        tasks = []
-        if acl_strategy is not None:
-            grants_to_migrate = self._gc.snapshot()
-            migrated_groups = self._group.snapshot()
-            principal_grants = self._principal_grants.get_interactive_cluster_grants()
+        all_tasks = []
+        if what == What.VIEW:
+            sequencer = TableMigrationSequencer(tables_to_migrate)
+            batches = sequencer.sequence_batches(what)
+            for batch in batches:
+                tasks = []
+                for view in batch:
+                    grants = self._compute_grants(
+                        view.src, acl_strategy, all_grants_to_migrate, all_migrated_groups, all_principal_grants
+                    )
+                    tasks.append(
+                        partial(
+                            self._migrate_view,
+                            view,
+                            grants,
+                        )
+                    )
+                Threads.strict("migrate views", tasks)
+                all_tasks.extend(tasks)
+                if len(batches) > 1:
+                    self._init_seen_tables()
         else:
-            acl_strategy = []
-        for table in tables_to_migrate:
-            grants = []
-            if what is not None and table.src.what != what:
-                continue
-            if AclMigrationWhat.LEGACY_TACL in acl_strategy:
-                grants.extend(self._match_grants(table.src, grants_to_migrate, migrated_groups))
-            if AclMigrationWhat.PRINCIPAL in acl_strategy:
-                grants.extend(self._match_grants(table.src, principal_grants, migrated_groups))
-            tasks.append(partial(self._migrate_table, table.src, table.rule, grants))
-        Threads.strict("migrate tables", tasks)
-
-    def _migrate_table(self, src_table: Table, rule: Rule, grants: list[Grant] | None = None):
-        if self._table_already_migrated(rule.as_uc_table_key):
-            logger.info(f"Table {src_table.key} already migrated to {rule.as_uc_table_key}")
+            tables_in_scope = filter(lambda t: t.src.what == what, tables_to_migrate)
+            tasks = []
+            for table in tables_in_scope:
+                grants = self._compute_grants(
+                    table.src, acl_strategy, all_grants_to_migrate, all_migrated_groups, all_principal_grants
+                )
+                tasks.append(
+                    partial(
+                        self._migrate_table,
+                        table,
+                        grants,
+                    )
+                )
+            Threads.strict("migrate views", tasks)
+            all_tasks = tasks
+        # the below is useful for testing
+        return all_tasks
+
+    def _compute_grants(
```

Comment:
can we get a separate PR for `_compute_grants` with good unit test coverage just for that?

Reply:
well that's not new code tbh, just moved out of `migrate_tables` which was doing too much stuff, and make it reusable
as I understand it, there's already significant grants related unit tests in `test_table_migrate.py`

Reply:
see `test_migrate_acls_should_produce_proper_queries` in `test_table_migrate.py`